sematic/ui/packages/common/src/hooks/runHooks.ts

```diff
@@ -0,0 +1,121 @@
+import { Filter, RunListPayload } from "@sematic/common/src/ApiContracts";
+import { useHttpClient } from "@sematic/common/src/hooks/httpHooks";
+import { useMemo, useState, useCallback, useEffect, useRef } from "react";
+import useAsyncFn from "react-use/lib/useAsyncFn";
+import { Run } from "src/Models";
+
+export type QueryParams = {[key: string]: string};
+export const PAGE_SIZE = 25;
+
+const defaultQueryParams = {};
+export function useFetchRunsFn(runFilters: Filter | undefined = undefined,
+    otherQueryParams: QueryParams = defaultQueryParams) {
+    const [isLoaded, setIsLoaded] = useState(false);
+
+    const queryParams = useMemo(() => {
+        let params = {...otherQueryParams};
+        if (!!runFilters) {
+            params.filters = JSON.stringify(runFilters)
+        }
+        return params;
+    }, [otherQueryParams, runFilters]);
+
+    const {fetch} = useHttpClient();
+
+    const [state, load] = useAsyncFn(async (overrideQueryParams: QueryParams = {}) => {
+        const finalQueryParams = {
+            ...queryParams,
+            ...overrideQueryParams
+        }
+        const qString = (new URLSearchParams(finalQueryParams)).toString();
+        const response = await fetch({
+            url: `/api/v1/runs?${qString}`
+        });
+        const payload: RunListPayload = await response.json();
+        setIsLoaded(true);
+        return payload;
+    }, [queryParams, fetch]);
+
+    const {loading: isLoading, error, value: runs} = state;
+
+    return {isLoaded, isLoading, error, runs: runs as RunListPayload, load};
+}
+
+export function useFetchRuns(runFilters: Filter | undefined = undefined,
+    otherQueryParams: {[key: string]: string} = defaultQueryParams) {
+    const {isLoaded, isLoading, error, runs, load} = useFetchRunsFn(runFilters, otherQueryParams);
+
+    const reloadRuns = useCallback(async () => {
+        const payload = await load();
+        return payload.content;
+    }, [load]);
+
+    useEffect(() => {
+        load();
+    }, [load])
+
+    return {isLoaded, isLoading, error, runs: runs?.content, reloadRuns};
+}
+
+export function useRunsPagination(runFilters: Filter | undefined = undefined,
+    otherQueryParams: {[key: string]: string} = defaultQueryParams) {
+    
+    const [page, setPage] = useState(0);
+
+    const [currentPageData, setCurrentPageData] = useState<Array<Run>>([]);
+
+    const pagesCache = useRef<Array<RunListPayload>>([]);
+
+    const queryParams = useMemo(() => {
+        return {
+            ...otherQueryParams,
+            limit: PAGE_SIZE.toString()
+        }
+    }, [otherQueryParams]);
+
+    const {isLoaded, isLoading, error, load} = useFetchRunsFn(runFilters, queryParams);
+
+    // Previous page should always be available in cache
+    const previousPage = useCallback(async () => {
+        const newPage = page - 1;
+        setCurrentPageData(pagesCache.current[newPage].content);
+        setPage(newPage);
+    }, [page]);
```

Comment:
Shouldn't this traversal use page-sticky atoms as well?
This way, when refreshing, you wouldn't be taken back to the first page, and you could share a specific page with other people as well.

Reply:
So far, our backend cannot handle random access to a specific page, data has to flow in using sequential cursor tokens. So implementing deep-linking to pages might not perform well. 

To pin down a specific run, I guess it is more useful to recommend user cites a run details page instead.

I am inclined to skip this feature for now.